### PR TITLE
Feature/add slacking model for training

### PR DESCRIPTION
**Pull request by duchoang**
- [x] Add slacking model, modify run_training.sh script
- [x] Ready for review

To run the training with slacking model:
```./run_training.sh -d dataset -o output -m slacking```

### CODE REVIEW

File: mlp/training/examples.py [hunk 1]
```diff
@@ -44,3 +44,10 @@ def shuffle(self):
         else:
             self.features = np.reshape(np.array(shuffled), (len(shuffled_labels),) + shuffled[0].shape)
             self.labels = np.reshape(np.array(shuffled_labels), (len(shuffled_labels)))
+
+    def print_statistic(self, example_name = "example", label_id_mapping = {}):
+        print "Statistic of ", example_name, ":"
```
Comment (tmbo) [inline mlp/training/examples.py]: Please use a logger instead.

File: mlp/training/acceleration_dataset.py [hunk 2]
```diff
@@ -170,8 +170,12 @@ def __init__(self, directory, test_directory=None, label_mapper=lambda x: x, add
             examples = self.load_examples(directory, label_mapper)
             examples.shuffle()
 
+            examples.print_statistic("all dataset", self.label_id_mapping)
+
             train, test = examples.split(self.TRAIN_RATIO)
-        
+
+        train.print_statistic("train", self.label_id_mapping)
+        test.print_statistic("test", self.label_id_mapping)
```
Comment (tmbo) [inline mlp/training/acceleration_dataset.py]: Can this be moved to the training script? As this is not necessary for the notebooks.

File: mlp/training/mlp_model.py [hunk 3]
```diff
@@ -50,38 +50,75 @@ def generate_default_model(self, num_labels):
...
         layers.append(Affine(
+            nout=num_labels,
+            init=init_norm,
+            bias=bias_init,
+            activation=Logistic()))
+
+        model = Model(layers=layers)
+        return model
+
+    def generate_slacking_model(self, num_labels):
```
Comment (tmbo) [inline mlp/training/mlp_model.py]: How about creating an extra class for the models? They are static anyway and don't really belong to the trainer.

**Comment by tmbo** [comment]
What are you trying to do with the label mapping? There is already a label mapping implementation in the `AccelerationDataset`. You can use it using something linke:

```python
exercises = ["biceps curls (left)", "biceps-curl", "bice", "arms/biceps-curl", "BC ", "bicep curls", "bicep",
            "arms/lateral-raise", "lateral raises", "lateral", "LR", "LR "
             "triceps-extension	", "tc ", "TE", "tc"
            ]

slacking = ["", "walking"]

def label_mapper(label):
    if label in slacking:
        return "-"
    elif label in exercises:
        return "E"
    else:
        return None

dataset = CSVAccelerationDataset('datasets/Exercise', label_mapper=label_mapper)
```

File: mlp/preprocess_data.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        ys = [label] * len(xs)
+        result.append(ExampleColl(xs, ys))
+    return result
+
+
+def main(dataset_directory, output_directory, ratio, is_slacking):
+    """Main entry point."""
+
+    # 1/ Load the dataset
+    dataset = load_data(dataset_directory, map_label)
```
Comment (duchoang) [inline mlp/preprocess_data.py]: @tomstockton actually the `load_data()` here is slightly the same with the method `AccelerationDataset.load_examples()`.
- In `load_examples` it returns the big object `ExampleColl` which contain all data. 
- But in here i want to generate the set of `ExampleColl`, each of them will correspond to only one same label. So in this case the size will be 4 (bicep, tricep, lateral, non-exercise).
- Then in the next step, I reuse the method `ExampleColl.split()` to divide the data for that label.

In addition, depend on the slacking model or exercise model, we will generate the dataset differently
1. For slacking model, convert all different exercises to just "exercise"
2. For exercise model, remove the ExampleColl with label non-exercise

Comment (tmbo) [inline mlp/preprocess_data.py]: Alright, let me rephrase: Why do you want to clone the dataset with your labels instead of doing the mapping on the fly?

Comment (duchoang) [inline mlp/preprocess_data.py]: Ok I got your point, at first I intend to separate 2 task::

1/ Map label, divide dataset:
```bash
./prepare_dataset.sh -d dataset -o output/ -r 80 -s
```
2/ Run training on folder output:
```bash
./run_training.sh -d output/train -t output/test/ -m -slacking
```

Then I want to take some trained input (in folder output/train/*.csv), put into the playground and see the classification result.

So do you think should we merge 2 steps? (maybe I only use this script locally for experiment)

Comment (tmbo) [inline mlp/preprocess_data.py]: In the future we will probably not split the dataset anymore. Instead we will curate a test & validate dataset with the data labeled as good as possible (even the slacking data using different labels for different activities) to be able to debug where the model has weaknesses.

**Comment by tmbo** [comment]
This PR only covers the very basic use case of training a model with a defined shape on more data. What we actually need is a way to train multiple models to evaluate them. Therefore, I still don't think this PR covers where we want to go for the following reasons:
- The automation trains only a single model. What we need would be a way to define lets say 10 models and hit the start button. And evaluate them after they are all done (manually).
- Model parameters scattered across code & files
- No support for non MLP models, e.g. convolutions
